Implement BIP 340-342 validation (Schnorr/taproot/tapscript)

sipa opened the pull request:
This is an implementation of the Schnorr/taproot consensus rules proposed by BIPs [340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki), [341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki), and [342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki).

See the list of commits [below](https://github.com/bitcoin/bitcoin/pull/19953#issuecomment-691815830). No signing or wallet support of any kind is included, as testing is done entirely through the Python test framework.

This is a successor to https://github.com/bitcoin/bitcoin/pull/17977 (see discussion following [this comment](https://github.com/bitcoin/bitcoin/pull/17977#issuecomment-682285983)), and will have further changes squashed/rebased. The history of this PR can be found in #19997.

sipa committed:
Add TaggedHash function (BIP 340)

This adds the TaggedHash function as defined by BIP340 to the hash module, which
is used in BIP340 and BIP341 to produce domain-separated hashes.

sipa committed:
refactor: keep spent outputs in PrecomputedTransactionData

A BIP-341 signature message may commit to the scriptPubKeys and amounts
of all spent outputs (including other ones than the input being signed
for spends), so keep them available to signature hashing code.

jl2012 committed:
Implement Taproot signature hashing (BIP 341)

This implements the new sighashing scheme from BIP341, with all relevant
whole-transaction values precomputed once and cached.

Includes changes to PrecomputedTransactionData by Pieter Wuille.

sipa committed:
scripted-diff: put ECDSA in name of signature functions

In preparation for adding Schnorr versions of `CheckSig`, `VerifySignature`, and
`ComputeEntry`, give them an ECDSA specific name.

-BEGIN VERIFY SCRIPT-
sed -i 's/CheckSig(/CheckECDSASignature(/g' $(git grep -l CheckSig ./src)
sed -i 's/VerifySignature(/VerifyECDSASignature(/g' $(git grep -l VerifySignature ./src)
sed -i 's/ComputeEntry(/ComputeEntryECDSA(/g' $(git grep -l ComputeEntry ./src)
-END VERIFY SCRIPT-

sipa committed:
Support for Schnorr signatures and integration in SignatureCheckers (BIP 340)

This enables the schnorrsig module in libsecp256k1, adds the relevant types
and functions to src/pubkey, as well as in higher-level `SignatureChecker`
classes. The (verification side of the) BIP340 test vectors is also added.

sipa committed:
refactor: rename scriptPubKey in VerifyWitnessProgram to exec_script

The old name is confusing, as it doesn't store a scriptPubKey, but the
actually executed script.

sipa committed:
Implement Taproot validation (BIP 341)

This includes key path spending and script path spending, but not the
Tapscript execution implementation (leaf 0xc0 remains unemcumbered in
this commit).

Includes constants for various aspects of the consensus rules suggested
by Jeremy Rubin.

sipa committed:
Use ScriptExecutionData to pass through annex hash

Instead of recomputing the annex hash every time a signature is verified, compute it
once and cache it in a new ScriptExecutionData structure.

jl2012 committed:
Implement Tapscript script validation rules (BIP 342)

This adds a new `SigVersion::TAPSCRIPT`, makes the necessary interpreter
changes to make it implement BIP342, and uses them for leaf version 0xc0
in Taproot script path spends.

sipa committed:
Make Taproot spends standard + policy limits

This adds a `TxoutType::WITNESS_V1_TAPROOT` for P2TR outputs, and permits spending
them in standardness rules. No corresponding `CTxDestination` is added for it,
as that isn't needed until we want wallet integration. The taproot validation flags
are also enabled for mempool transactions, and standardness rules are added
(stack item size limit, no annexes).

sipa committed:
Activate Taproot/Tapscript on regtest (BIP 341, BIP 342)

Define a versionbits-based activation for the new consensus rules on regtest.
No activation or activation mechanism is defined for testnet or mainnet.

sipa committed:
tests: add BIP340 Schnorr signature support to test framework

Add a pure Python implementation of BIP340 signing and verification, tested against
the BIP's test vectors.

sipa commented:
Here is a list of all the commits in topological order:

* initial implementation (https://github.com/sipa/bitcoin/compare/ba4b3fbcf21317a9410f3be913e4437d58968091...6c1edf0d73deada863b8bf32879d70831c291633)
  * 3430eaec88 Add TaggedHash function (BIP 340)
  * ba285d8d2b Keep spent outputs in PrecomputedTransactionData during validation
  * b63f19a4bd Implement Taproot signature hashing (BIP 341)
  * 466b0229ca refactor: put ECDSA in name of signature function
  * 76f2e3e14d Support for Schnorr signatures and integration in SignatureCheckers (BIP 340)
  * ddfc3657d0 refactor: rename scriptPubKey in VerifyWitnessProgram to exec_script
  * 804ff644f4 Implement Taproot validation (BIP 341)
  * d384d67873 Use ScriptExecutionData to pass through annex hash
  * a73d5e1d8b Implement Tapscript script validation rules (BIP 342)
  * a7ab59707c Make Taproot spends standard + policy limits
  * 24fd22f2ef Activate Taproot/Tapscript on regtest (BIP 341, BIP 342)
  * 89a14b5b7d [TESTS] Add BIP340 Schnorr signature support to test framework
  * 6c1edf0d73 [TESTS] Functional tests for Schnorr/Taproot/Tapscript

```
PREV="$(git rev-parse HEAD)"; (git log --oneline upstream/master..HEAD; echo "$(git rev-parse upstream/master) --- [TAPROOT] begin: initial implementation ---") | while read C L; do if [ "d${L:0:14}" == "d--- [TAPROOT] " ]; then if [ "d$PREV" != "" ]; then echo "* ${L:21:-4} (https://github.com/sipa/bitcoin/compare/$C...$PREV)"; fi; PREV=$C; PREVL=$L; else echo "  * $C $L"; fi; done | tac
```

sipa committed:
tests: functional tests for Schnorr/Taproot/Tapscript

A large functional test is added that automatically generates random transactions which
exercise various aspects of the new rules, and verifies they are accepted into the mempool
(when appropriate), and correctly accepted/rejected in (Python-constructed) blocks.

Includes sighashing code and many tests by Johnson Lau.
Includes a test by Matthew Zipkin.
Includes several tests and improvements by Greg Sanders.

jnewbery reviewed (commented):
Concept ACK :)

Perhaps the pure refactor commits can be split into their own PR to reduce the size of this?

jnewbery commented on configure.ac:
```diff
@@ -1645,7 +1645,7 @@ if test x$need_bundled_univalue = xyes; then
 fi
 
-ac_configure_args="${ac_configure_args} --disable-shared --with-pic --enable-benchmark=no --with-bignum=no --enable-module-recovery"
+ac_configure_args="${ac_configure_args} --disable-shared --with-pic --enable-benchmark=no --with-bignum=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental"
```
Does it make sense to remove the requirement for the `-enable-experimental` flag to build the schnorrsig module?

fanquake replied:
This was [discussed in the previous PR](https://github.com/bitcoin/bitcoin/pull/17977/files#r475990462).

luke-jr
> Should this be non-experimental before merging?

gmaxwell
> I don't think so: It probably shouldn't be marked non-experimental until after it deployed for activation in Bitcoin because it wouldn't be good to encourage third party users of it while it is still easy to make incompatible changes in Bitcoin (e.g. as was just done. :) )

sipa
> Or, maybe more likely: the API in libsecp256k1 may change still significantly in the near future (e.g. support for variable-length messages, batch validation, ...) even after we treat the scheme itself final.

sipa commented:
@jnewbery There are only two of them. The variable name one is very small, and the ECDSA naming one isn't really a standalone useful change. I think that one could be changed into a scripted-diff though.

flack commented on src/policy/policy.cpp:
```diff
@@ -237,6 +239,36 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+        if (witnessversion == 1 && witnessprogram.size() == WITNESS_V1_TAPROOT_SIZE && !p2sh) {
+            // Taproot spend (non-P2SH-wrapped, version 1, witness program size 32; see BIP 341)
+            auto stack = MakeSpan(tx.vin[i].scriptWitness.stack);
+            if (stack.size() >= 2 && !stack.back().empty() && stack.back()[0] == ANNEX_TAG) {
```
maybe move that block down into the one below, then you don't have to check `stack.size() >= 2` twice

sipa replied:
I would rather not, because if later changes do add supported annexes, it'll need to revert to this flow anyway (where the branch dealing with annexes pops it from the stack, and the next if looks at the stack size after removing it).

instagibbs commented:
concept ACK, just confirming for now this PR is identical to the old PR #17977

sipa committed:
--- [TAPROOT] Refactors ---

sipa committed:
--- [TAPROOT] BIP340/341/342 consensus rules ---

sipa committed:
--- [TAPROOT] Tests ---

sipa committed:
--- [TAPROOT] Regtest activation and policy ---

sipa commented:
Reordered commits a bit, replace the ECDSA naming one with a scripted diff, and organized the commits in sections. The end state is identical to what it was before./

benthecarman commented:
Could the first 2 commits be done as separate PRs? That should slightly reduce the size of this one

sipa commented:
@benthecarman That was suggested earlier by @jnewbery. I think splitting off 28 trivial lines of refactors from a 2400-line PR isn't going to change much.

sipa committed:
tests: add generic qa-asset-based script verification unit test

This adds a unit test that does generic script verification tests,
with positive/negative witnesses/scriptsigs, under various flags.
The test data is large (several MB) so it's stored in the qa-assets
repo.

sipa commented:
I added a unit test too now, with test vectors that were extracted from 20000 runs of the feature_taproot.py test (with the largest tests removed), minimized using libfuzzer's coverage tracking to 105.

jnewbery commented:
I was surprised to learn that this was a 2500-line PR. By directory:

- `/test/functional` - 1790 lines
- `/src/test` - 134 lines
- `/src` (exc `test`) - 817 lines

So the majority of code in this PR is tests (which is a good thing!)

I agree with sipa that it's not necessary to split off the first two commits (especially now that they're separated to the top of the branch). Equally no harm in doing so if that'd reduce the workload.

instagibbs commented:
No code changes aside from unit test commit https://github.com/bitcoin/bitcoin/pull/19953/commits/9673fd9999a7fef4bfc5b074fa1861938042df10

benthecarman commented on src/pubkey.h:
```diff
@@ -206,6 +207,24 @@ class CPubKey
+
+    /** Verify a 64-byte Schnorr signature.
+     *
+     * If the signature is not exactly 64 bytes, false is returned.
```
`VerifySchnorr` has `assert(sigbytes.size() == 64);` so false won't be returned. Should either change this comment or change the assertion to `if (sigbytes.size() != 64) return false;`

sipa replied:
Fixed.

benthecarman commented on src/script/script.cpp:
```diff
@@ -328,3 +331,11 @@ bool GetScriptOp(CScriptBase::const_iterator& pc, CScriptBase::const_iterator en
+
+bool IsOpSuccess(const opcodetype& opcode)
+{
+    return (opcode == 0x50 || opcode == 0x62 || opcode == 0x89 ||
+            opcode == 0x8a || opcode == 0x8d || opcode == 0x8e ||
+            (opcode >= 0x7e && opcode <= 0x81) || (opcode >= 0x83 && opcode <= 0x86) ||
+            (opcode >= 0x95 && opcode <= 0x99) || (opcode >= 0xbb && opcode <= 0xfe));
```
nit: would be easier to check if this was correct if they were ordered the same as the BIP

sipa replied:
Done. I intentionally didn't change the implementation in test_framework, as this provides weak additional evidence that the semantics didn't change compared to the older code.

benthecarman commented on src/script/script_error.cpp:
```diff
@@ -75,6 +75,10 @@ std::string ScriptErrorString(const ScriptError serror)
         case SCRIPT_ERR_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION:
             return "Taproot version reserved for soft-fork upgrades";
+        case SCRIPT_ERR_DISCOURAGE_OP_SUCCESS:
+            return "SUCCESSx reserved for soft-fork upgrades";
```
nit: when referring to other op codes in the error messages they are normally prefixed with `OP_`, to be consistent this could be changed to `OP_SUCCESSx reserved for...`

sipa replied:
Done.

jaybny commented:
#### My expectation was to do a code review for taproot, but unfortunately, I cannot ACK the concept or the approach of this PR at this time. 

After initial research and deep dive, I ask the following:  
1. Will a majority of bitcoin users benefit from this PR?  
_for consensus changes this answer should be a 'YES'_
2. Does this PR create new security concerns or attack vectors for the typical bitcoin user?
_for consensus changes this answer should be a 'NO'_

**Below are my concerns with questions regarding this PR, answers and clarifications will hopefully bring me around.**

### Background 
Original motivation to review this PR came from a tweet with a bounty. I am honored to help bitcoin, and am reviewing this PR independant of monetary expectations.  

Following documents have been read and reviewed in detail specifically for this review, skipping some code and proofs.
- [x] [PR Review Club for 17977](https://bitcoincore.reviews/17977-3)
- [x] [[bitcoin-dev] Taproot](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-January/015614.html)
- [x] [Homomorphic Payment Addresses and the Pay-to-Contract Protocol](https://arxiv.org/pdf/1212.3257.pdf)
- [x] [BIP340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki), [BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki), [BIP342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki)
- [x] [Simple Schnorr Multi-Signatures with Applications to Bitcoin](https://eprint.iacr.org/2018/068.pdf)
- [x] [Impact of Taproot and Schnorr on Address Clustering Analysis of Bitcoin](http://blockchain.cs.ucl.ac.uk/wp-content/uploads/2020/04/UCL_CBT_DiscussionPaper_Q12020_Anania_2020.pdf) 

Reason for not a simple "concept NACK" is because of the known information gap on my end. Specifically, I have not read most of bitcoin bips since early segwit. 
* My knowledge and use of bitcoin has been almost exclusively with P2PKH, including custom coded raw transactions on the utxo level. 
* Have not yet personally implemented BIP32 - Hierarchical Deterministic Wallets, nor SegWit 

### Questions
some of these are leading some of these are clarifying 
1. Are my current ECDSA private/public key pairs compatible with Schnorr?
2. Where are the specific steps of transforming keys to Schnorr public/private key pairs? 
3. is there any difference in key management between ECDSA pairs and Schnorr pairs?
4. is Schnorr compatible with BIP32? 
5. If we define _complexity risk_ as a measure of required knowledge in relation to security expectations, does this PR add complexity risk? 
6. Can we measure the _complexity risk_ delta of this PR?
7. Would an expert ECDSA pk pair security manager need more knowledge to become an expert Schnorr pk pair manager? 
8. Would a future expert Schnorr pk pair manger already be an expert ECDSA pk pair manager? 
9. If we define a _hard-knowledge-fork_ as needing more knowledge and a _soft-knowledge-fork_ as needing less knowledge to use bitcoin, is this Taproot PR a _hard-knowledge-fork_ or _soft-knowledge-fork_? 
10. Can Taproot be implemented with ECDSA? [Technical: Pay-to-contract and Sign-to-contract](https://www.reddit.com/r/Bitcoin/comments/d3lffo/technical_paytocontract_and_signtocontract/)?
11. Is Taproot available for ECDSA in this PR? 

### Concerns
**MultiSig**
I’m concerned that we are explicitly promoting MultiSig as a net positive for securing users private keys, and as far as i can tell, the consensus is that multisig comes down to a single leader in practice. 

[poll: Is the use of multisig on a bitcoin transaction, a security feature or a false sense of security? ](https://twitter.com/jaybny/status/1308297471735418882)

**Key Reuse** 
I’m concerned that we are implicitly assuming that key reuse is wrong, but as far as i can tell, the consensus is that code reuse is fine for most cases.  

**Highest concern is if the efficiency improvements, like removal of double sha256 hashing and others, are assuming no key reuse, creating a new vector for user error, since a majority of bitcoin users reuse their keys.**


similar sentiments:
> "While the key holders are expected not to reuse key pair, little could be done to stop payers to reuse an address. Unfortunately, key-pair reuse has been a social and technical norm since the creation of Bitcoin (the first tx made in block 170 reused the previous public key). I don’t see any hope to change this norm any time soon, if possible at all." [bitcoin-dev](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-December/016549.html)

> > Overall, we are left with concerns both about the merit of doing Taproot
> > versus alternatives, as well as the process through which we got to be here.
> > 1) Is Taproot actually more private than bare MAST and Schnorr separately? What
> > are the actual anonymity set benefits compared to doing the separately?
> 
> Yes, presuming single-pubkey-single-signature remains a common authorisation pattern. 
[bitcoin-dev](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-February/017622.html)

and fud: 
> [Similarly, the SIGHASH_SINGLE bug for example would have been disastrous for this scheme. In general, the Blockchain this is used in must not allow spending more than one output with a single 
> signature.](https://github.com/ElementsProject/scriptless-scripts/blob/master/md/partially-blind-swap.md)

red flags:
> [Homomorphic Payment Addresses and the Pay-to-Contract Protocol](https://arxiv.org/pdf/1212.3257.pdf)
> * Customer public key “a" in the tx cannot be used twice for same Public-Key of merchant 
> * signaling variation? address(a,c) vs address(P,c) - 
>     * (a,c) becomes a private key 

other:
DISCOURAGE_FLAG - why the need for this flag? can’t we say - ON_POLICY_CHANGE(Transform the mempool)?  

flagged for followup:
[BIP340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki)
> This complicates the implementation of secure signing protocols in scenarios in which a group of mutually distrusting signers work together to produce a single joint signature 

> Using the first option would be slightly more efficient for verification (around 10%), but we prioritize compactness, and therefore choose option 3.
batch verification

> breaking an X only key - at most a small constant faster

> nonce - leaking secret

> every public key - has two corresponding secret keys

> not using the same private key - across different signing schemes - signatures can leak secret key through nonce reuse

> open up new nonce attack vectors

> However, if this optimization is used and additionally the signature verification at the end of the signing algorithm is dropped for increased efficiency, signers must ensure the public key is correctly calculated and not taken from untrusted sources.

> While recent academic papers claim that they are also possible with ECDSA, consensus support for Schnorr signature verification would significantly simplify the constructions.”

[BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)
> Combining all these ideas in a single proposal would be an extensive change, be hard to review, and likely miss new discoveries that otherwise could have been made along the way.

> Just like other existing output types, taproot outputs should never reuse keys, for privacy reasons. This does not only apply to the particular leaf that was used to spend an output but to all leaves committed to in the output. If leaves were reused, it could happen that spending a different output would reuse the same Merkle branches in the Merkle proof. 

> fixes the verification capabilities of offline signing devices by including amount and scriptPubKey in the signature message, avoids unnecessary hashing

> Hashes that go into the signature message and the message itself are now computed with a single SHA256 invocation instead of double SHA256. There is no expected security improvement by doubling SHA256 because this only protects against length-extension attacks against SHA256 which are not a concern for signature messages because there is no secret data. Therefore doubling SHA256 is a waste of resources.

> The public key is directly included in the output in contrast to typical earlier constructions which store a hash of the public key or script in the output. This has the same cost for senders and is more space efficient overall if the key-based spending path is taken. [2]


[BIP342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki)
> The inefficient OP_CHECKMULTISIG and OP_CHECKMULTISIGVERIFY opcodes are disabled. Instead, a new opcode OP_CHECKSIGADD is introduced to allow creating the same multisignature policies in a batch-verifiable way. 

> Disabled script opcodes The following script opcodes are disabled in tapscript: OP_CHECKMULTISIG and OP_CHECKMULTISIGVERIFY[2]. The disabled opcodes behave in the same way as OP_RETURN, by failing and terminating the script immediately when executed, and being ignored when found in unexecuted branch of the script.

> This way, new SIGHASH modes can be added, as well as NOINPUT-tagged public keys and a public key constant which is replaced by the taproot internal key for signature validation.

> Why is a limit on script size no longer needed? Since there is no scriptCode directly included in the signature hash (only indirectly through a precomputable tapleaf hash), the CPU time spent on a signature check is no longer proportional to the size of the script being executed.

### PR Purpose
as stated 

> privacy, efficiency, and flexibility of Bitcoin's scripting capabilities

> This proposal aims to improve privacy, efficiency, and flexibility of Bitcoin's scripting capabilities without adding new security assumptions[1]. Specifically, it seeks to minimize how much information about the spendability conditions of a transaction output is revealed on chain at creation or spending time and to add a number of upgrade mechanisms, while fixing a few minor but long-standing issues

Taproot is an optional feature, and P2PKH key reusers, not concerned with privacy, do not need to us it. Yet, they still get the benefits of smaller multisig transactions.  

**Is this [chart](https://eprint.iacr.org/2018/068.pdf) the only benefits of this PR for the majority of users? Is it worth consensus soft-fork? What are the risks?** 
![image](https://user-images.githubusercontent.com/777326/93874364-19fd3d80-fc88-11ea-8f65-c478398f0108.png)
## Taproot ##
> Taproot's advantages become apparent under the assumption that most applications involve outputs that could be spent by all parties agreeing 

This use of practical reality built into the bitcoin protocol is exactly what we need more off! Maybe this abstraction can one day be made between alice and bob, because many times alice is really bob in a dress, and bob has no double spend risk on the tx from bob to bob.  

**defaut spending path**
from what i understand, since in practice over 85% of bitcoin transaction are single key, the default path will be basically p2pkh (or the Segwit + Schnorr equivalent)   

**user story**
If Taproot tree path spending is not for the 85% of bitcoin users, then who is it for? actual user stories will go a long way here.

sipa commented:
@jaybny All of those questions are about the BIPs themselves, and not about the code changes to implement them, and thus out of scope here. I suggest you discuss them on the bitcoin-dev mailing list, https://bitcoin.stackexchange.com, or other media.

I'll give my answers for your list of short questions, but if you have follow up discussion, please take it elsewhere. It would be distracting to code reviewers here.

1. Yes, intentionally. This is explained in BIP340.
2. You remove the first byte of a 33-byte compressed pubkey to obtain a 32-byte x-only pubkey.
3. No.
4. Yes.
5. BIP341/342 increase the functionality of the scripting language, which also implies there are more ways someone can screw up when doing nontrivial things. On the other hand, certain weaknesses that used to exist in legacy scripts and segwit scripts are fixed.
6. That sounds subjective.
7. No.
8. Not sure what you mean.
9. Of course these are softforks.
10. In theory yes, but doing most interesting things with it requires a way to have a single tweaked key that can be spent using consent of multiple parties. This is a *lot* harder with ECDSA than with Schnorr.
11. This PR implements the BIP341 and BIP342 consensus rules. If those BIPs would have specified support for Taproot with ECDSA, then this PR would have needed to comply with that. BIP341 uses Schnorr signatures in its key spending path, and BIP342 explicitly specifies that all opcodes involving signatures use Schnorr signatures (OP_CHECKSIG, OP_CHECKSIGVERIFY, OP_CHECKSIGADD) or are disables (OP_CHECKMULTISIG, OP_CHECKMULTISIGVERIFY).

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -86,13 +105,13 @@ def is_x_coord(self, x):
 
     def lift_x(self, x):
-        """Given an X coordinate on the curve, return a corresponding affine point."""
+        """Given an X coordinate on the curve, return a corresponding affine point for which the y-coordinate is even."""
```
inconsistent capitalization and hyphenation

sipa replied:
Fixed.

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +393,161 @@ def sign_ecdsa(self, msg, low_s=True):
+    return (P[0].to_bytes(32, 'big'), not SECP256K1.has_even_y(P))
+
+def tweak_add_privkey(key, tweak, negated=False):
+    """Tweak a private key (after optionally negating it)."""
```
unclear whether it will be negated again if it is already negated, or if it will be negated if it is not already negated

sipa replied:
Changed the interface to not take a negated argument, but infer it from the key. This should be cleaner and simpler.

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +393,161 @@ def sign_ecdsa(self, msg, low_s=True):
+    """Create a Schnorr signature (see BIP 340)."""
+
+    if aux is None:
+        aux = bytes(0 for _ in range(32))
```
simpler to just use `bytes(32)`. python wouldn't let you have uninitialized memory.

sipa replied:
TIL. Done.

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +393,161 @@ def sign_ecdsa(self, msg, low_s=True):
+        num_tests = 0
+        with open(os.path.join(sys.path[0], 'test_framework', 'bip340_test_vectors.csv'), newline='', encoding='utf8') as csvfile:
+            reader = csv.reader(csvfile)
+            reader.__next__()
```
why call magic methods directly when you can use `next(reader)`? arguably more readable.

sipa replied:
This was copied from the BIP340 reference code. You may want to update it there too.

Done.

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +393,161 @@ def sign_ecdsa(self, msg, low_s=True):
+    if aux is None:
+        aux = bytes(0 for _ in range(32))
+
+    assert(len(key) == 32)
```
parens unnecessary

sipa replied:
Done, in many places.

ysangkok commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +393,161 @@ def sign_ecdsa(self, msg, low_s=True):
+def tweak_add_pubkey(key, tweak):
+    """Tweak a public key and return whether the result was negated."""
+
+    assert(len(key) == 32)
```
parens unnecessary

sipa replied:
Done in many places.

sipa commented:
I addressed comments by @benthecarman and @ysangkok. See "Updates 2020-09-22" in the new PR #19997 for the changes that were applied.

instagibbs commented:
reviewed updates to https://github.com/bitcoin/bitcoin/pull/19953/commits/fe5cf2ed6adbf2271553f97c61de4d59bee20c05

`IsOpSuccess` was compared against the BIP directly rather than previous iteration, which I had hand-calculated painstakingly. It matches the BIP.

instagibbs commented on src/test/script_tests.cpp:
```diff
@@ -1610,5 +1640,50 @@ BOOST_AUTO_TEST_CASE(bitcoinconsensus_verify_script_invalid_flags)
+        SCRIPT_VERIFY_CHECKSEQUENCEVERIFY |
+        SCRIPT_VERIFY_WITNESS |
+        SCRIPT_VERIFY_NULLFAIL |
+        SCRIPT_VERIFY_TAPROOT;
```
Why no `MINIMALIF`?

instagibbs replied:
... because it's not used for consensus, along with CLEANSTACK et al. Right.

sipa replied:
Added a comment in script/interpreter.h.

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/19953/commits/fe5cf2ed6adbf2271553f97c61de4d59bee20c05

Reviewed all commits one by one, side by side with my latest known-good review. Poked at new unit tests/mutated logic as a sanity check.

One request I'd have is some documentation on how to extract the unit test, if possible. Maybe it was hacked together, but still having a general description of using the fuzzing construct could be helpful for people to look for smaller/better sets. For example, someone could create coverage for MINIMALIF. (I did a mutation test for it and it passed)

benthecarman commented on src/script/interpreter.cpp:
```diff
@@ -381,6 +439,9 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
     static const valtype vchTrue(1, 1);
 
+    // sigversion cannot be TAPROOT here, as it admits no script execution.
+    assert(sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0 || sigversion == SigVersion::TAPSCRIPT);
```
If a new sig version is added in the future, we probably want it to fail here.

instagibbs replied:
eh ok, guess crashing would be best in this case

sipa replied:
Yeah, I'm intentionally only using equality checks with sigversion (all over interpreter.cpp), so that there are no surprises when a new one is added.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1375,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+        ss << (CHashWriter(SER_GETHASH, 0) << witstack->back()).GetSHA256();
+    }
+
+    // Data about the output(s)
```
Fixed.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1565,7 +1675,7 @@ bool GenericTransactionSignatureChecker<T>::CheckSchnorrSignature(Span<const uns
     }
     uint256 sighash;
-    if (!SignatureHashSchnorr(sighash, execdata, *txTo, nIn, hashtype, sigversion, this->txdata)) return false;
+    if (!SignatureHashSchnorr(sighash, execdata, *txTo, nIn, hashtype, sigversion, 0x00, this->txdata)) return false;
```
Done.

sipa commented:
I addressed comments by @instagibbs, and moved the unit test vectors to be qa-assets based (see corresponding PR https://github.com/bitcoin-core/qa-assets/pull/27). Changes are in two new sections in #19997.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,23 +1291,76 @@ uint256 GetOutputsSHA256(const T& txTo)
+                m_spent_outputs[inpos].scriptPubKey[0] == OP_1) {
+                // Treat every native witness v1 spend as a Taproot spend. This only works if spent_outputs was
+                // provided as well, but if it wasn't, actual validation will fail anyway.
+                uses_bip341_taproot = true;
```
No, one input can be segwit and another can be taproot.

benthecarman replied:
This should just be an optimization, ie if there are 100 inputs and the first is segwit v0 and the second is segwit v1, with @promag's suggestion it would then stop checking with both `uses_bip341_taproot` and `uses_bip143_segwit` set correctly, vs with the current implementation it would continue and check the remaining 98 inputs.

sipa replied:
Oh, I misread. Indeed, that's a potentially useful and simple optimization. I'll implement it when I make further changes.

sipa replied:
Done.

instagibbs commented:
re-ACK https://github.com/bitcoin/bitcoin/pull/19953/commits/528ed883558b0bdac97fa64ada560906c322d021

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1375,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+        assert(false);
+    }
+    assert(in_pos < tx_to.vin.size());
+    assert(cache != nullptr && cache->m_bip341_taproot_ready && cache->m_spent_outputs_ready);
```
Done.

sipa commented:
I addressed comments by @promag, added a functional test to verify CLEANSTACK is consensus-enforced in tapscript, and gave tapscript's MINIMALIF a separate error code (effectively addressing #20009 to the extent it affects this PR).

benthecarman reviewed (approved):
ACK `80efcba`

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/80efcba5e7ad84f68f5cd59c05fc87134fc9fb69

ysangkok commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1389,176 @@ uint256 GetOutputsSHA256(const T& txTo)
+    // Hash type
+    const uint8_t output_type = (hash_type == SIGHASH_DEFAULT) ? SIGHASH_ALL : (hash_type & SIGHASH_OUTPUT_MASK); // Default (no sighash byte) is equivalent to SIGHASH_ALL
+    const uint8_t input_type = hash_type & SIGHASH_INPUT_MASK;
+    if (!(hash_type <= 0x03 || (hash_type >= 0x81 && hash_type <= 0x83))) return false;
```
This line is very compact.

Looks like those numeric literals are based on SIGHASH_INPUT_MASK. If they are, could they not be constructed from it? It is weird to have logic based partly on the enum, and partly on hard coded numeric literals.

I think this could be split up into two if's, each returning false. Then, each of those branches could be explaining the situation and the reason for rejection.

Every time I see a negation of a boolean operator, I do de-morgan in my head. Either way, if you prefer this with or without the outer negation, I think splitting it up would really be preferable, it would be readable for all kinds of brains.

This is not complicated logic, and it is easy to avoid having it look complex. Just my two cents :)

sipa replied:
That would be incorrect.

de-morganing it gives you `!(hash_type <= 0x03) && !(hash_type >= 0x81 && hash_type <= 0x83)`, which is a conjunction at the top level. Splitting that up into two separate `if`s would make it a disjunction.

sipa committed:
tests: dumping and minimizing of script assets data

This adds a --dumptests flag to the feature_taproot.py test, to dump all its
generated test cases to files, in a format compatible with the
script_assets_test unit test. A fuzzer for said format is added as well, whose
primary purpose is coverage-based minimization of those dumps.

sipa commented:
A few updates (see https://github.com/bitcoin/bitcoin/pull/19997 for detailed history):
* Fixed/clarified a few code comments addressing anonymous review
* A few generic improvements to the functional test (include ECDSA tests with uncompressed pubkeys, cache witnesses)
* Added support to the functional test to dump its generated tests in json format compatible with the unit test, and a fuzz "test" to help minimize it (it actually works as a fuzz test itself now too, though probably not a very good one)
* Rebased on master to adapt the test for #20006.

decryp2kanon commented on src/chainparams.cpp:
```diff
@@ -86,6 +86,11 @@ class CMainParams : public CChainParams {
 
+        // Deployment of Taproot (BIPs 340-342)
+        consensus.vDeployments[Consensus::DEPLOYMENT_TAPROOT].bit = 2;
+        consensus.vDeployments[Consensus::DEPLOYMENT_TAPROOT].nStartTime = 1199145601; // January 1, 2008
```
just curious, why January 1, 2008?

instagibbs replied:
Copying `DEPLOYMENT_TESTDUMMY` I suspect

sipa replied:
Indeed, just whatever time sufficiently in the past.

benthecarman commented:
reACK `c849dae`

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/c849dae8ce05ef5c92df6ec1bb77b9ea2e38240a

gruve-p commented:
Ack https://github.com/bitcoin/bitcoin/pull/19953/commits/c849dae8ce05ef5c92df6ec1bb77b9ea2e38240a

decryp2kanon commented on src/test/script_tests.cpp:
```diff
@@ -1470,6 +1500,7 @@ BOOST_AUTO_TEST_CASE(script_HasValidOps)
     BOOST_CHECK(!script.HasValidOps());
 }
 
+
```
nit: a new line?

sipa replied:
Fixed.

decryp2kanon commented on test/functional/test_framework/key.py:
```diff
@@ -1,15 +1,30 @@
-# Copyright (c) 2019 Pieter Wuille
+# Copyright (c) 2019-2020 Pieter Wuille
+
```
nit: new line?

sipa replied:
Fixed.

promag commented on src/script/interpreter.h:
```diff
@@ -227,15 +237,15 @@ class GenericTransactionSignatureChecker : public BaseSignatureChecker
 using MutableTransactionSignatureChecker = GenericTransactionSignatureChecker<CMutableTransaction>;
 
-bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = nullptr);
+bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = nullptr, ScriptExecutionData execdata = {});
```
6577f465f3dad2460b2bee02ca662c7ebbe72e4c
nit, avoids copying `ScriptExecutionData` and error is still the last one.
```cpp
bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, const ScriptExecutionData& execdata, ScriptError* error = nullptr);
bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = nullptr)
{
    ScriptExecutionData execdata;
    EvalScript(stack, script, flags, checker, sigversion, execdata, error);
}

sipa replied:
That won't work, `EvalScript` needs a mutable `ScriptExecutionData` object.

It would be slightly better to split the execdata into a fixed-at-VerifyScript-time part and a changed-during-execution part, but that'd be a even more annoying interface change.

sipa replied:
Actually, it's possible to avoid the copy still by just making it mutable, and using your idea of a wrapper to provide a default:

https://github.com/sipa/bitcoin/commit/09b24d503a7b9baace0be3f51e53f2a095cf160b

sipa replied:
I've incorporated this patch, as I think it conceptually makes sense: `ScriptExecutionData`'s contents has a linear history, so there is no need to pass it as `const`.

sipa commented:
Addressed comments by @decryp2kanon and @promag. See detailed history in #19988.

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/7dfb7239773e7974bdb61f648bf8351a808a63a5

benthecarman commented:
reACK 7dfb723

jonatack reviewed (commented):
First pass of code review and debug build of each commit up to 3c2d549ee. No blocking issues seen so far. A few minor comments follow; feel free to pick/choose/ignore. Continuing on the second half of the commits.

Reviewers: if helpful, there are several review club sessions on this BIP340-342 implementation at https://bitcoincore.reviews/meetings-components/#consensus

Edit: a new dedicated `taproot` review club section has been added:

https://bitcoincore.reviews/meetings-components/#taproot

jonatack commented on src/script/script.cpp:
```diff
@@ -328,3 +331,11 @@ bool GetScriptOp(CScriptBase::const_iterator& pc, CScriptBase::const_iterator en
+    return opcode == 80 || opcode == 98 || (opcode >= 126 && opcode <= 129) ||
+           (opcode >= 131 && opcode <= 134) || (opcode >= 137 && opcode <= 138) ||
+           (opcode >= 141 && opcode <= 142) || (opcode >= 149 && opcode <= 153) ||
+           (opcode >= 187 && opcode <= 254);
```
332dc13b maybe simplify
```diff
     return opcode == 80 || opcode == 98 || (opcode >= 126 && opcode <= 129) ||
-           (opcode >= 131 && opcode <= 134) || (opcode >= 137 && opcode <= 138) ||
-           (opcode >= 141 && opcode <= 142) || (opcode >= 149 && opcode <= 153) ||
+           (opcode >= 131 && opcode <= 134) || opcode == 137 || opcode == 138 ||
+           opcode == 141 || opcode == 142 || (opcode >= 149 && opcode <= 153) ||
            (opcode >= 187 && opcode <= 254);
```

instagibbs replied:
IIRC the text as-is completely "matches" the BIP text, meaning if it gives a range here, it gives a range in the BIP text. I think as-is is easier to pattern match.

jonatack replied:
True, though the BIP prose here is inconsistent (see the first 2 elements) and could be updated. The code could opt for the simpler version. Not a biggie tho :)

sipa replied:
I prefer to keep it exactly as in the BIP for now, but no reason the BIP can't be updated to be more clear.

jonatack reviewed (commented):
Checkpoint after review of commits 6e42c8d0721a9 and 332dc13b, cross-referencing the implementation with the BIPs. The code appears to follow the spec where I have verified. Continuing forward with the remaining commits.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -381,6 +438,9 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
     static const valtype vchTrue(1, 1);
 
+    // sigversion cannot be TAPROOT here, as it admits no script execution.
+    assert(sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0 || sigversion == SigVersion::TAPSCRIPT);
```
I'd rather be explicit in every calling site about which sigversions are affected.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1394,178 @@ uint256 GetOutputsSHA256(const T& txTo)
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata, const T& tx_to, const uint32_t in_pos, const uint8_t hash_type, const SigVersion sigversion, const uint8_t key_version, const PrecomputedTransactionData& cache)
+{
+    uint8_t ext_flag;
```
I believe that not initializing a variable is better than initializing it incorrectly. The former lets you use tooling (ubsan, valgrind) to detect invalid usage, instead of creating a potentially very hard to discover bug (I don't think it matters much here, as it's a very simple case, but as a matter of principle I dislike initialization of variables just to avoid a risk of them being uninitialized).

sipa commented on src/validation.cpp:
```diff
@@ -1538,14 +1538,20 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState &state, const C
+        spent_outputs.reserve(tx.vin.size());
+
+        for (const auto& txin : tx.vin) {
+            const COutPoint &prevout = txin.prevout;
```
Done.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1389,178 @@ uint256 GetOutputsSHA256(const T& txTo)
 
 template <class T>
-void PrecomputedTransactionData::Init(const T& txTo)
+void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut> spent_outputs)
```
Done.

sipa commented on src/script/script.h:
```diff
@@ -44,6 +44,17 @@ static const unsigned int LOCKTIME_THRESHOLD = 500000000; // Tue Nov  5 00:53:20
+// Tag for input annex. If there are at least two witness elements for a transaction input,
+// and the first byte of the last element is 0x50, this last element is called annex, and
+// has meanings independent of the script
+static const unsigned int ANNEX_TAG = 0x50;
```
Done.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1389,178 @@ uint256 GetOutputsSHA256(const T& txTo)
+static const CHashWriter HASHER_TAPTWEAK = TaggedHash("TapTweak");
+
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata, const T& tx_to, const uint32_t in_pos, const uint8_t hash_type, const SigVersion sigversion, const uint8_t key_version, const PrecomputedTransactionData& cache)
```
Done.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1389,178 @@ uint256 GetOutputsSHA256(const T& txTo)
-template void PrecomputedTransactionData::Init(const CTransaction& txTo);
-template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo);
+template void PrecomputedTransactionData::Init(const CTransaction& txTo, std::vector<CTxOut> spent_outputs);
+template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo, std::vector<CTxOut> spent_outputs);
```
Done.

sipa commented on src/script/interpreter.h:
```diff
@@ -114,41 +125,111 @@ enum
 
     template <class T>
-    void Init(const T& tx);
+    void Init(const T& tx, std::vector<CTxOut> spent_outputs);
```
Done.

sipa commented on src/pubkey.h:
```diff
@@ -206,6 +207,26 @@ class CPubKey
+
+public:
+    /** Construct an x-only pubkey from exactly 32 bytes. */
+    XOnlyPubKey(Span<const unsigned char> input);
```
Renamed them both to `bytes`.

sipa commented on src/script/sigcache.cpp:
```diff
@@ -34,18 +35,30 @@ class CSignatureCache
+        // just write our 32-byte entropy, and then pad with 'E' for ECDSA and
+        // 'S' for Schnorr (followed by 0 bytes).
+        static const unsigned char PADDING_ECDSA[32] = {'E'};
+        static const unsigned char PADDING_SCHNORR[32] = {'S'};
```
Done.

sipa commented on src/pubkey.cpp:
```diff
@@ -166,6 +167,26 @@ int ecdsa_signature_parse_der_lax(const secp256k1_context* ctx, secp256k1_ecdsa_
+    return secp256k1_schnorrsig_verify(secp256k1_context_verify, sigbytes.data(), msg.begin(), &pubkey);
+}
+
+bool XOnlyPubKey::CheckPayToContract(const XOnlyPubKey& base, const uint256& hash, const bool parity) const
```
Done.

sipa commented on src/script/interpreter.h:
```diff
@@ -114,41 +125,111 @@ enum
+    /** Whether m_validation_weight_left is initialized. */
+    bool m_validation_weight_left_init = false;
+    /** How much validation weight is left (decremented for every successful non-empty signature check). */
+    int64_t m_validation_weight_left;
```
Done.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -568,6 +634,13 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                         valtype& vch = stacktop(-1);
+                        // Tapscript requires minimal IF/NOTIF inputs as a consensus rule.
+                        if (sigversion == SigVersion::TAPSCRIPT) {
+                            if (vch.size() > 1 || (vch.size() == 1 && vch[0] != 1)) {
```
Done.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1291,35 +1394,178 @@ uint256 GetOutputsSHA256(const T& txTo)
+    if (sigversion == SigVersion::TAPSCRIPT) {
+        assert(execdata.m_tapleaf_hash_init);
+        ss << execdata.m_tapleaf_hash;
+        assert(key_version == 0); // key_version must be 0 for now
```
Done.

sipa commented:
I addressed comments by @jonatack. See detailed history in #19997.

benthecarman commented:
reACK 07dd29e

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/07dd29e1b50b8fcba91c36386ed5267a2e60c240

sipa commented:
I made a few more changes related to signature checking:

* The `key_version` (currently always 0) is no longer passed as an argument into `SignatureHashSchnorr`, because if a new one was introduced, it would need a new SigVersion as well. So instead use the latter to derive the `key_version`. @ariard This partially reverts an earlier comment you had, but I think the comments make it clearer still.
* Introduce distinct script validation errors for the various ways Schnorr signatures (in Taproot/Tapscript) can fail (invalid length, invalid sighash type, or actual Schnorr validation failure). This allows more accurate testing in feature_taproot.py, but did require passing down `ScriptError` in a few more places.

As always, details of the changes are in #19953.

benthecarman commented:
reACK 4a7e017

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/4a7e0171c0eec437c5de2b0eb316c5088e5b07b0

gruve-p commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/4a7e0171c0eec437c5de2b0eb316c5088e5b07b0

fjahr reviewed (commented):
Commit 87dad64920eeaf41e5e10c503f1e0f48c5d120d4 does not have a description at the moment and it looks like you might have wanted to squash it?

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1708,11 +1731,41 @@ static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion,
+            if (!VerifyTaprootCommitment(control, program, exec_script)) {
+                return set_error(serror, SCRIPT_ERR_WITNESS_PROGRAM_MISMATCH);
+            }
+            if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION) {
```
This is no longer the case after later commits, where actual script semantics are added.

kallewoof reviewed (commented):
Tested ACK, code review up to and including regtest section (5ffd2ee94fcce072842365fe96e29169ac3c0524)

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+    }
+
+    // Data about the input/prevout being spent
+    const auto* witstack = &tx_to.vin[in_pos].scriptWitness.stack;
```
Fixed.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1403 @@
+        random.shuffle(scripts)
+        tap = taproot_construct(pubs[0], scripts)
+        add_spender(spenders, "opsuccess/bare", standard=False, tap=tap, leaf="bare_success", failure={"leaf": "bare_nop"}, **ERR_CLEANSTACK)
+        add_spender(spenders, "opsuccess/exexecif", standard=False, tap=tap, leaf="unexecif_success", failure={"leaf": "unexecif_nop"}, **ERR_CLEANSTACK)
```
Done.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1403 @@
+            # Add the 50 highest-value inputs
+            unspents = node.listunspent()
+            random.shuffle(unspents)
+            unspents.sort(key=lambda x: -int(x["amount"] * 100000000))
```
Nice, done.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1403 @@
+        """
+
+        # Generate some coins to fund the wallet.
+        node.generate(10)
```
Indded, fixed.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1403 @@
+# - lists of values
+# - callables which, when fed the context object as argument, produce any of these
+#
+# # The DEFAULT_CONTEXT object specifies a standard signing process, with many overridable knobs.
```
Fixed.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -371,6 +368,66 @@ static bool EvalChecksig(const valtype& vchSig, const valtype& vchPubKey, CScrip
+/** Helper for OP_CHECKSIG, OP_CHECKSIGVERIFY, and (in Tapscript) OP_CHECKSIGADD.
+ *
+ * A return value of false means the script fails entirely. When true is returned, the
+ * fSuccess variable indicates whether the signature check itself succeeded.
```
Fixed.

sipa commented on src/validation.cpp:
```diff
@@ -1539,13 +1539,19 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState &state, const C
+        txdata.Init(tx, std::move(spent_outputs));
     }
 
     for (unsigned int i = 0; i < tx.vin.size(); i++) {
```
Added.

sipa commented:
Addressed comments by @fjahr and @kallewoof above. Details are in #19997.

kallewoof commented:
Re-ACK sans the fuccess typo.

Local diff between 4a7e0171c0eec437c5de2b0eb316c5088e5b07b0 and cc036c251c16eda1cdc558db2049e8db0f3a50b7 is

<details><summary>Diff</summary>
<p>

```diff
diff --git a/src/pubkey.cpp b/src/pubkey.cpp
index 689328765..4d734fc89 100644
--- a/src/pubkey.cpp
+++ b/src/pubkey.cpp
@@ -173,7 +173,8 @@ XOnlyPubKey::XOnlyPubKey(Span<const unsigned char> bytes)
     std::copy(bytes.begin(), bytes.end(), m_keydata.begin());
 }
 
-bool XOnlyPubKey::VerifySchnorr(const uint256& msg, Span<const unsigned char> sigbytes) const {
+bool XOnlyPubKey::VerifySchnorr(const uint256& msg, Span<const unsigned char> sigbytes) const
+{
     assert(sigbytes.size() == 64);
     secp256k1_xonly_pubkey pubkey;
     if (!secp256k1_xonly_pubkey_parse(secp256k1_context_verify, &pubkey, m_keydata.data())) return false;
diff --git a/src/pubkey.h b/src/pubkey.h
index 65f3fdf38..0f784b86e 100644
--- a/src/pubkey.h
+++ b/src/pubkey.h
@@ -207,7 +207,8 @@ public:
     bool Derive(CPubKey& pubkeyChild, ChainCode &ccChild, unsigned int nChild, const ChainCode& cc) const;
 };
 
-class XOnlyPubKey {
+class XOnlyPubKey
+{
 private:
     uint256 m_keydata;
 
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 92994a56e..1067a5293 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -392,13 +392,13 @@ static bool EvalChecksigTapscript(const valtype& sig, const valtype& pubkey, Scr
         return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
     } else if (pubkey.size() == 32) {
         if (success && !checker.CheckSchnorrSignature(sig, pubkey, sigversion, execdata, serror)) {
-            return false;
+            return false; // serror is set
         }
     } else {
         /*
          *  New public key version softforks should be defined before this `else` block.
          *  Generally, the new code should not do anything but failing the script execution. To avoid
-         *  consensus bugs, it should not modify any existing values (including success).
+         *  consensus bugs, it should not modify any existing values (including `success`).
          */
         if ((flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_PUBKEYTYPE) != 0) {
             return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_PUBKEYTYPE);
@@ -411,7 +411,7 @@ static bool EvalChecksigTapscript(const valtype& sig, const valtype& pubkey, Scr
 /** Helper for OP_CHECKSIG, OP_CHECKSIGVERIFY, and (in Tapscript) OP_CHECKSIGADD.
  *
  * A return value of false means the script fails entirely. When true is returned, the
- * fSuccess variable indicates whether the signature check itself succeeded.
+ * fuccess variable indicates whether the signature check itself succeeded.
  */
 static bool EvalChecksig(const valtype& sig, const valtype& pubkey, CScript::const_iterator pbegincodehash, CScript::const_iterator pend, ScriptExecutionData& execdata, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* serror, bool& success)
 {
@@ -1814,9 +1814,7 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
         }
 
         // Tapscript enforces initial stack size limits (altstack is empty here)
-        if (stack.size() > MAX_STACK_SIZE) {
-            return set_error(serror, SCRIPT_ERR_STACK_SIZE);
-        }
+        if (stack.size() > MAX_STACK_SIZE) return set_error(serror, SCRIPT_ERR_STACK_SIZE);
     }
 
     // Disallow stack item size > MAX_SCRIPT_ELEMENT_SIZE in witness stack
@@ -1900,7 +1898,7 @@ static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion,
         if (stack.size() == 1) {
             // Key path spending (stack size is 1 after removing optional annex)
             if (!checker.CheckSchnorrSignature(stack.front(), program, SigVersion::TAPROOT, execdata, serror)) {
-                return false;
+                return false; // serror is set
             }
             return set_success(serror);
         } else {
diff --git a/src/validation.cpp b/src/validation.cpp
index fdabff9bb..0b78ba677 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1550,6 +1550,7 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState &state, const C
         }
         txdata.Init(tx, std::move(spent_outputs));
     }
+    assert(txdata.m_spent_outputs.size() == tx.vin.size());
 
     for (unsigned int i = 0; i < tx.vin.size(); i++) {
 
diff --git a/test/functional/feature_taproot.py b/test/functional/feature_taproot.py
index 82ccdae06..7b534c1c2 100755
--- a/test/functional/feature_taproot.py
+++ b/test/functional/feature_taproot.py
@@ -100,7 +100,7 @@ import random
 # - lists of values
 # - callables which, when fed the context object as argument, produce any of these
 #
-# # The DEFAULT_CONTEXT object specifies a standard signing process, with many overridable knobs.
+# The DEFAULT_CONTEXT object specifies a standard signing process, with many overridable knobs.
 #
 # The get(ctx, name) function can evaluate a name, and cache its result in the context.
 # getter(name) can be used to construct a callable that evaluates name. For example:
@@ -1070,7 +1070,7 @@ def spenders_taproot_active():
         random.shuffle(scripts)
         tap = taproot_construct(pubs[0], scripts)
         add_spender(spenders, "opsuccess/bare", standard=False, tap=tap, leaf="bare_success", failure={"leaf": "bare_nop"}, **ERR_CLEANSTACK)
-        add_spender(spenders, "opsuccess/exexecif", standard=False, tap=tap, leaf="unexecif_success", failure={"leaf": "unexecif_nop"}, **ERR_CLEANSTACK)
+        add_spender(spenders, "opsuccess/unexecif", standard=False, tap=tap, leaf="unexecif_success", failure={"leaf": "unexecif_nop"}, **ERR_CLEANSTACK)
         add_spender(spenders, "opsuccess/return", standard=False, tap=tap, leaf="return_success", failure={"leaf": "return_nop"}, **ERR_OP_RETURN)
         add_spender(spenders, "opsuccess/undecodable", standard=False, tap=tap, leaf="undecodable_success", failure={"leaf": "undecodable_nop"}, **ERR_UNDECODABLE)
         add_spender(spenders, "opsuccess/undecodable_bypass", standard=False, tap=tap, leaf="undecodable_success", failure={"leaf": "undecodable_bypassed_success"}, **ERR_UNDECODABLE)
@@ -1245,9 +1245,6 @@ class TaprootTest(BitcoinTestFramework):
         of all valid inputs, except one invalid one.
         """
 
-        # Generate some coins to fund the wallet.
-        node.generate(10)
-
         # Construct a bunch of sPKs that send coins back to the host wallet
         self.log.info("- Constructing addresses for returning coins")
         host_spks = []
@@ -1284,7 +1281,7 @@ class TaprootTest(BitcoinTestFramework):
             # Add the 50 highest-value inputs
             unspents = node.listunspent()
             random.shuffle(unspents)
-            unspents.sort(key=lambda x: -int(x["amount"] * 100000000))
+            unspents.sort(key=lambda x: int(x["amount"] * 100000000), reverse=True)
             if len(unspents) > 50:
                 unspents = unspents[:50]
             random.shuffle(unspents)

```
</p>
</details>

(but don't take my word for it).

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -363,15 +360,75 @@ static bool EvalChecksig(const valtype& vchSig, const valtype& vchPubKey, CScrip
+/** Helper for OP_CHECKSIG, OP_CHECKSIGVERIFY, and (in Tapscript) OP_CHECKSIGADD.
+ *
+ * A return value of false means the script fails entirely. When true is returned, the
+ * fuccess variable indicates whether the signature check itself succeeded.
```
```suggestion
 * success variable indicates whether the signature check itself succeeded.
```

sipa replied:
Done.

instagibbs reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/cc036c251c16eda1cdc558db2049e8db0f3a50b7

benthecarman commented:
reACK cc036c2

achow101 reviewed (commented):
light ACK cc036c251c16eda1cdc558db2049e8db0f3a50b7

Skimmed the tests, will review them in further detail soon.

fjahr commented:
Code review ACK cc036c251c16eda1cdc558db2049e8db0f3a50b7

Hope I can get enough testing in soon to get to a full ACK.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+static const CHashWriter HASHER_TAPSIGHASH = TaggedHash("TapSighash");
+
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const T& tx_to, uint32_t in_pos, uint8_t hash_type, SigVersion sigversion, uint8_t key_version, const PrecomputedTransactionData& cache)
```
Fixed.

sipa commented on src/script/script.h:
```diff
@@ -47,7 +47,7 @@ static const uint32_t LOCKTIME_MAX = 0xFFFFFFFFU;
 // and the first byte of the last element is 0x50, this last element is called annex, and
 // has meanings independent of the script
-static const unsigned int ANNEX_TAG = 0x50;
+static constexpr unsigned int ANNEX_TAG = 0x50;
```
Fixed.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1504,58 +1791,146 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
 }
 
-static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion, const std::vector<unsigned char>& program, unsigned int flags, const BaseSignatureChecker& checker, ScriptError* serror)
+static bool VerifyTaprootCommitment(const std::vector<unsigned char>& control, const std::vector<unsigned char>& program, const CScript& script, uint256* tapleaf_hash)
```
Fixed.

sipa commented:
I addressed comments by @jnewbery, @instagibbs, and @achow101 above. The changes to the final tree are listed in #19997.

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/19953/commits/0e2a5e448f426219a6464b9aaadcc715534114e6

benthecarman commented:
reACK 0e2a5e4

kallewoof commented:
reACK 0e2a5e448f426219a6464b9aaadcc715534114e6

jonasnick commented:
ACK 0e2a5e448f426219a6464b9aaadcc715534114e6 almost only looked at bip340/libsecp related code

achow101 reviewed (commented):
ACK 0e2a5e448f426219a6464b9aaadcc715534114e6

fjahr commented:
reACK 0e2a5e448f426219a6464b9aaadcc715534114e6

sipa commented on test/functional/test_framework/key.py:
```diff
@@ -223,7 +242,7 @@ def set(self, data):
                 # valid y
-                if (p[1] & 1) != (data[0] & 1):
+                if data[0] & 1:
                     p = SECP256K1.negate(p)
```
I think the only thing that's wrong here is the comment: with this change, it's not longer "correcting" the oddness; it's just negating if an odd Y coordinate is desired.

The code is necessary though, but possibly untested. It's what constructs a point from a compressed public key. It's only used for ECDSA (as BIP340 public keys are x-only, not compressed), and unused in the current tests (which only use the signing side of ECDSA).

Going to leave this for a follow-up, as it's not directly related to Taproot testing.

fanquake replied:
This comment is also being addressed in #20161.

sipa commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +392,161 @@ def sign_ecdsa(self, msg, low_s=True):
+                        sig_actual = sign_schnorr(seckey, msg, aux_rand)
+                        self.assertEqual(sig.hex(), sig_actual.hex(), "BIP340 test vector %i (%s): sig mismatch" % (i, comment))
+                    except RuntimeError as e:
+                        self.assertFalse("BIP340 test vector %i (%s): signing raised exception %s" % (i, comment, e))
```
Indeed, will fix in a follow-up.

fanquake replied:
This is being addressed in #20161.

jamesob commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+    const uint8_t spend_type = (ext_flag << 1) + (have_annex ? 1 : 0); // The low bit indicates whether an annex is present.
+    ss << spend_type;
+    if (input_type == SIGHASH_ANYONECANPAY) {
+        ss << tx_to.vin[in_pos].prevout;
```
Note to reviewers: serializes as `[hash][out_index]` per `COutPoint` ([c.f. BIP](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message)).

jamesob commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+    ss << spend_type;
+    if (input_type == SIGHASH_ANYONECANPAY) {
+        ss << tx_to.vin[in_pos].prevout;
+        ss << cache.m_spent_outputs[in_pos];
```
Note to reviewers: serializes as `[amount i.e. nValue][scriptPubKey]` per `CTxOut` ([c.f. BIP](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message)).

jamesob commented on src/script/interpreter.cpp:
```diff
@@ -1322,6 +1378,75 @@ template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo,
+        ss << in_pos;
+    }
+    if (have_annex) {
+        ss << (CHashWriter(SER_GETHASH, 0) << witstack.back()).GetSHA256();
```
So far as I can tell, the [ordering of serialization](https://github.com/jamesob/bitcoin/blob/301959fa7482cf749504864789e6f9f4470b8bb6/src/serialize.h#L238) is reversed here relative to what [the BIP](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message) says ("the SHA256 of *(compact_size(size of annex) || annex)*"). Existing serialization looks like it has the size coming after the annex itself. Even if I'm right, not sure how much this matters.

sipa replied:
I don't understand. Why would the length go after the data? Nothing serializes that way, and the functional tests would fail if that was the case.

jamesob replied:
Based on the serialization code I link to, it looks as though the size is being written after the data itself (which is the reverse of what the BIP says), but maybe I'm missing something?

sipa replied:
That serialization code you link to:
* Is serializing a `Span<unsigned char>` as a fixed length object, without any length at all. It's invoking the `Stream` function `write(const unsigned char* data, size_t length)`, which writes the `length` bytes starting at `data`. It doesn't write the length itself.
* Isn't used here (witstack.back() is a `std::vector<unsigned char>`, not a `Span), which has different serialization code a bit further down the file.

jamesob replied:
Hm yeah I guess I must be wrong because the tests encode in [the right order](https://github.com/bitcoin/bitcoin/pull/19953/commits/f06e6d03452cf5e0b1a0863afb08c9e6d3ef452e#diff-3f25b0e0fbc5dd36a62417827b38517c9c5bdbfc3f4787d5925527c344bda104R775) (per `ser_string`'s [definition](https://github.com/jamesob/bitcoin/blob/301959fa7482cf749504864789e6f9f4470b8bb6/test/functional/test_framework/messages.py#L106-L107)). Maybe I'm looking at the wrong serializer.

jamesob reviewed (commented):
Partial review; more soon.

- [x] f8c099e220 --- [TAPROOT] Refactors ---
- [x] 107b57df9f scripted-diff: put ECDSA in name of signature functions
- [x] 8bd2b4e784 refactor: rename scriptPubKey in VerifyWitnessProgram to exec_script
- [x] 5d62e3a68b refactor: keep spent outputs in PrecomputedTransactionData
- [x] 450d2b2371 --- [TAPROOT] BIP340/341/342 consensus rules ---
- [x] 9eb590894f Add TaggedHash function (BIP 340)
- [x] 5de246ca81 Implement Taproot signature hashing (BIP 341)

ajtowns reviewed (commented):
ACK 0e2a5e448f426219a6464b9aaadcc715534114e6 - code review, just nits. However signet activation params are missing (should be disabled as per mainnet/testnet)

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1679,14 +1682,35 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
 }
 
-static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion, const std::vector<unsigned char>& program, unsigned int flags, const BaseSignatureChecker& checker, ScriptError* serror)
+static bool VerifyTaprootCommitment(const std::vector<unsigned char>& control, const std::vector<unsigned char>& program, const CScript& script)
```
Without #13062 that's annoying to do, as it means constructing both a `valtype` and a `CScript` with the same data. You're right that future version leafs may not want a script at all, but until then, little reason to add this complication.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1679,14 +1682,35 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
+
+static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion, const std::vector<unsigned char>& program, unsigned int flags, const BaseSignatureChecker& checker, ScriptError* serror, bool is_p2sh)
+{
+    CScript exec_script; //!< Actually executed script (last stack item in P2WSH; implied P2PKH script in P2WPKH; leaf script in P2TR)
```
Yeah, not sure that's worth changing without other improvements though.

hebasto commented:
This regression introduced in 4567ba034c5ae6e6cc161360f7425c9e844738f0 is fixed in #20180.

ariard commented on src/script/interpreter.cpp:
```diff
@@ -363,15 +360,75 @@ static bool EvalChecksig(const valtype& vchSig, const valtype& vchPubKey, CScrip
+        /*
+         *  New public key version softforks should be defined before this `else` block.
+         *  Generally, the new code should not do anything but failing the script execution. To avoid
+         *  consensus bugs, it should not modify any existing values (including `success`).
```
Can we enforce this assign-once property with either some cpp magic or compiler option ? I've no idea.

sipa replied:
I'm sure there are ways to solve these softforkability guarantees more generically by encapsulating modifiable properties in an object... but the risks from refactoring consensus code to allow that are probably far bigger than the risk a bug would be missed in future consensus logic (probably a very rare event).

ariard reviewed (commented):
Code Review ACK 0e2a5e4. Mainly reviewed change since my last review in #17977 at 84ec870.

One thing I'm still inquiring is scope of test coverage. I've started to look at it only now and it's still trying to map if every spec object is correctly covered. I'll try if I can come with any comment improvement suggestion.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -363,15 +360,75 @@ static bool EvalChecksig(const valtype& vchSig, const valtype& vchPubKey, CScrip
+    success = !sig.empty();
+    if (success) {
+        // Implement the sigops/witnesssize ratio test.
+        // Passing with an upgradable public key version is also counted.
```
They can certainly define their own cost rules, as long as the cost is at least 50 vbytes per check. I'm not sure what you mean by "without branching".

ariard replied:
I meaned that we do the sigops/witnesssize ratio test before the pubkey size one. If a future softforked new pubkey type comes with its own new ratio test, maybe the code structure isn't going to be adequate ?

sipa commented on test/functional/test_framework/script.py:
```diff
@@ -738,3 +741,113 @@ def test_cscriptnum_encoding(self):
+             - a function, which specifies how to compute the hashing partner
+               in function of the hash of whatever it is combined with
+
+    Returns: script (sPK or redeemScript), tweak, {name:(script, leaf version, negation flag, innerkey, merklepath), ...}
```
See #20207. All of this was just outdated, thanks for noticing.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1458 @@
+    * failure: a dict of entries to override in the context when intentionally failing to spend (if None, no_fail will be set)
+    * standard: whether the (valid version of) spending is expected to be standard
+    * err_msg: a string with an expected error message for failure (or None, if not cared about)
+    * sigops_weight: the pre-taproot sigops weight consumed by a successful spend
```
Added in #20207.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1504,58 +1791,144 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
 {
-    CScript scriptPubKey;
+    const int path_len = (control.size() - TAPROOT_CONTROL_BASE_SIZE) / TAPROOT_CONTROL_NODE_SIZE;
+    const XOnlyPubKey p{uint256(std::vector<unsigned char>(control.begin() + 1, control.begin() + TAPROOT_CONTROL_BASE_SIZE))};
```
I added a bunch of comments around this in #20207. The first byte contains both the leaf version and the parity bit, btw.

sipa commented on src/pubkey.h:
```diff
@@ -169,7 +170,7 @@ class CPubKey
      * Check syntactic correctness.
      *
-     * Note that this is consensus critical as CheckSig() calls it!
+     * Note that this is consensus critical as CheckECDSASignature() calls it!
```
Added comments in #20207.

kiminuo replied:
https://github.com/bitcoin/bitcoin/blob/b2f5c38333fd2ec03fd1299a1bce507b25753f45/src/pubkey.h#L253 - is the asterisk in `*this` intentional?

MarcoFalke commented on src/Makefile.test.include:
```diff
@@ -1082,6 +1083,12 @@ test_fuzz_script_interpreter_LDADD = $(FUZZ_SUITE_LD_COMMON)
+test_fuzz_script_assets_test_minimizer_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
+test_fuzz_script_assets_test_minimizer_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
+test_fuzz_script_assets_test_minimizer_LDADD = $(FUZZ_SUITE_LD_COMMON)
+test_fuzz_script_assets_test_minimizer_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
```
nit: FUZZ_SUITE_LDFLAGS_COMMON

maflcko reviewed (commented):
Concept ACK 0e2a5e448f

left some questions in the test commits

MarcoFalke commented on src/test/script_tests.cpp:
```diff
@@ -1610,5 +1640,104 @@ BOOST_AUTO_TEST_CASE(bitcoinconsensus_verify_script_invalid_flags)
+            // a subset of test_flags.
+            if (fin || ((flags & test_flags) == flags)) {
+                bool ret = VerifyScript(tx.vin[idx].scriptSig, prevouts[idx].scriptPubKey, &tx.vin[idx].scriptWitness, flags, txcheck, nullptr);
+                BOOST_CHECK(ret);
```
Can you explain what the point of this unit test is? It seems that the python test is used to generate inputs for the fuzz test, which is used to minimize the corpus. The resulting corpus is fed into the unit test.

However, the unit test and fuzz test do the same thing. If they didn't, the fuzz test couldn't be used to capture the coverage.

I think the unit test (all changes in this commit) can be reverted if the fuzz asserts the `ret` return code directly.

sipa replied:
If the fuzz tests would assert the return code directly, they wouldn't be usable in fuzzing mode, and likely cause people to file bug reports if they'd try to run them. As is, the fuzz tests are just for assessing coverage, and can be used either as test directly, or used to minimize a corpus generated externally (with known validity/invalidity). Only the latter can be used as a unit test.

So I think the differences are:
* The unit test actually tests the script validation results against known valid/invalid things.
* The unit test can be run without building in fuzz mode.

MarcoFalke replied:
> with known validity/invalidity

Oh, I missed that the validity flag is included in the seed itself. Makes sense now.

> The unit test can be run without building in fuzz mode.

Heh, I am pretty sure the developers that have the seed dir cloned, run the unit tests, but not the fuzz tests is an empty set.

sipa commented on src/policy/policy.cpp:
```diff
@@ -206,6 +206,7 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
         // get the scriptPubKey corresponding to this input:
         CScript prevScript = prev.scriptPubKey;
 
+        bool p2sh = false;
```
Personal style, I guess. I don't think there is much of an objective difference.

MarcoFalke replied:
Overall less verbose and one less LOC, but obviously a style question. Resolving.

sipa commented on src/policy/policy.cpp:
```diff
@@ -237,6 +239,36 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
             }
         }
+
+        // Check policy limits for Taproot spends:
```
See follow-up in https://github.com/bitcoin/bitcoin/pull/20207.

MarcoFalke replied:
Which commit? I can't seem to find the one that touches validation.cpp

sipa replied:
Oops, done for real now.

MarcoFalke replied:
heh, thanks!

sipa commented on test/functional/test_framework/key.py:
```diff
@@ -369,3 +392,161 @@ def sign_ecdsa(self, msg, low_s=True):
+    def test_schnorr_testvectors(self):
+        """Implement the BIP340 test vectors (read from bip340_test_vectors.csv)."""
+        num_tests = 0
+        with open(os.path.join(sys.path[0], 'test_framework', 'bip340_test_vectors.csv'), newline='', encoding='utf8') as csvfile:
```
Ok, I wasn't aware of that mode being supported. Added in #20207.

sipa commented on test/functional/test_framework/key.py:
Done.

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1411 @@
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+        # Node 0 has Taproot inactive, Node 1 active.
+        self.extra_args = [["-whitelist=127.0.0.1", "-par=1", "-vbparams=taproot:1:1"], ["-whitelist=127.0.0.1", "-par=1"]]
```
I'm not sure, it may have been needed in an old version of this code. It doesn't seem needed anymore; addressed in #20207.

MarcoFalke replied:
One side-effect is that it speeds up tx-relay, but tx relay isn't tested in this script, so seems good to remove

sipa commented on test/functional/feature_taproot.py:
```diff
@@ -0,0 +1,1411 @@
+        self.log.info("  - Done")
+
+    def run_test(self):
+        self.connect_nodes(0, 1)
```
I probably copied it from somewhere ¯\_(ツ)_/¯

dgpv commented on test/functional/test_framework/script.py:
```diff
@@ -738,3 +741,113 @@ def test_cscriptnum_encoding(self):
+        if input_index < len(txTo.vout):
+            ss += sha256(txTo.vout[input_index].serialize())
+        else:
+            ss += bytes(0 for _ in range(32))
```
Post-merge comment/question after stumbling upon this while looking at the test framework code:

Is using 32 zero bytes instead of correct BIP341 behaviour (i.e. failing) intentional here ?

The code does not seem accidental, as it is not copied from legacy sighash function, and if I stick `assert 0` here, the test/functional/feature_taproot.py test fails, but it is not clear from the backtrace where exactly, a lot of `deep_eval()` in that backtrace.

If this is intentional, I think that a comment with explanation was in order here, to avoid confusion for the readers of the code.

If this was not intentional, then even if does not cause problems now, it might result in incorrect test behaviour in the future

dgpv replied:
@sipa, can you please comment on this ?

ajtowns replied:
I think it's testing that a SIGHASH_SINGLE signature without a corresponding output will fail, even if you generate an otherwise reasonable looking signature. If you change it to `bytes(42 for _ in range(32))` rather than an `assert 0`, the tests will still pass, demonstrating the exact value used isn't important. (If you change the length to something other than 32, you'll get an assertion failure slightly later when the length of hashed data is tested)

See the `need_vin_vout_mismatch` argument to `make_spender()`, which is set to `True` in the "Test SIGHASH_SINGLE behavior in combination with mismatching outputs" section. If you disable those two spenders, then adding the `assert 0` when there isn't a corresponding output works fine.

dgpv replied:
It is very much looks like you are correct, these two tests depend on this behavior of `TaprootSignatureHash`.

Without any comment explaining this, It will confuse people who read only a part of the code. Some people might copy that code and have problems later. I only cross-referenced the code with C++ implementation in Core while writing my own implementation in python, but I think it is definitely a possibility that someone may just copy the whole `TaprootSignatureHash`, and maybe even without looking at the comments.

Might it be better to just do `add_spender( ... , failure={"hashtype_actual": hashtype, "sighash": lambda(ctx): b"\x00"*32}, ... )` for these two tests ? It is not that `TaprootSignatureHash` is tested here, AFAIU, and expected behavior is incorrect hash returned, so why not just return it right away ?
